core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,405 @@
+    // worker running that task K times. Where K is more than the maxTaskFailures. Since the worker
+    // is notified of the decommissioning, the task failures should be treated as benign and not
+    // the job.
+
+    sc = createSparkContext(appConf)
+    val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+    val taskIdsKilled = new ConcurrentHashMap[Long, Boolean]
+    val listener = new RootStageAwareListener {
+      override def handleRootTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        val taskInfo = taskStart.taskInfo
+        if (taskInfo.index == 0) {
+          if (taskIdsKilled.size() < numTimesToKillWorkers) {
+            val workerInfo = executorIdToWorkerInfo(taskInfo.executorId)
+            decommissionWorkerOnMaster(workerInfo, "partition 0 must die")
+            killWorkerAfterTimeout(workerInfo, 1)
+            taskIdsKilled.put(taskInfo.taskId, true)
+          }
+        } else {
+          assert(false, s"Unknown task index ${taskInfo.index}")
+        }
+      }
+
+      override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+        val taskInfo = taskEnd.taskInfo
+        if (taskInfo.index == 0) {
+          // If a task has been killed then it shouldn't be successful
+          assert(taskInfo.successful === !taskIdsKilled.get(taskInfo.taskId))
+        } else {
+          assert(false, s"Unknown task index ${taskInfo.index}")
+        }
+      }
+    }
+    sc.addSparkListener(listener)
+    // single task job
+    val jobResult = sc.parallelize(1 to 1, 1).map(_ => {
+      Thread.sleep(5 * 1000L); 1
+    }).count()
+    assert(jobResult === 1)
+  }
+
+  test("decommission workers ensure that shuffle output is regenerated even with shuffle service") {
+    val conf = appConf
+    conf.set(config.Tests.TEST_NO_STAGE_RETRY, true)
+    conf.set(config.SHUFFLE_MANAGER, "sort")
+    conf.set(config.SHUFFLE_SERVICE_ENABLED, true)
+    makeWorkers(2)
+    val ss = new ExternalShuffleServiceHolder(conf)
+    sc = createSparkContext(conf)
+
+    // Here we will create a 2 stage job: The first stage will have two tasks and the second stage
+    // will have one task. The two tasks in the first stage will be long and short. We decommission
+    // and kill the worker after the short task is done. Eventually the driver should get the
+    // executor lost signal for the short task executor. This should trigger regenerating
+    // the shuffle output since we cleanly decommissioned the executor, despite running with an
+    // external shuffle service.
+    try {
+      val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+      val task0Killed = new AtomicBoolean(false)
+      // single task job
+      val listener = new RootStageAwareListener {
+        override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+          val taskInfo = taskEnd.taskInfo
+          if (taskInfo.index == 0) {
```

Comment:
The same as before (move the assert here and rewrite it). 

Here it is more important as the block scopes are too deep so move all the `assert`s up where the `else` only contains assertions only.

Reply:
oh I see that comment I am referring to is lost let me add it again